File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/MessageGeneral.java
package com.github.dedis.popstellar.model.network.method.message;

import androidx.annotation.NonNull;

import com.github.dedis.popstellar.model.Immutable;
import com.github.dedis.popstellar.model.network.method.message.data.Data;
import com.github.dedis.popstellar.model.network.method.message.data.message.WitnessMessageSignature;
import com.github.dedis.popstellar.model.objects.security.*;
import com.google.gson.Gson;

import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.util.*;

import timber.log.Timber;

/**
 * Container of a high level message.
 *
 * <p>It is encapsulated inside low level messages
 */
@Immutable
public final class MessageGeneral {

  private static final String TAG = MessageGeneral.class.getSimpleName();
  private static final MessageGeneral EMPTY =
      new MessageGeneral(null, null, null, null, null, new ArrayList<>());

  private final PublicKey sender;
  private final Base64URLData dataBuf;
  private final Data data;
  private final MessageID messageId;

  private Signature signature;
  private List<PublicKeySignaturePair> witnessSignatures = new ArrayList<>();

  public static MessageGeneral emptyMessage() {
    return EMPTY;
  }

  public MessageGeneral(
      PublicKey sender,
      Base64URLData dataBuf,
      Data data,
      Signature signature,
      MessageID messageID,
      List<PublicKeySignaturePair> witnessSignatures) {
    this.sender = sender;
    this.dataBuf = dataBuf;
    this.data = data;
    this.messageId = messageID;
    this.signature = signature;
    this.witnessSignatures = new ArrayList<>(witnessSignatures);
  }

  public MessageGeneral(KeyPair keyPair, Data data, Gson gson) {
    sender = keyPair.getPublicKey();
    this.data = data;
    String dataJson = gson.toJson(data, Data.class);
    Timber.tag(TAG).d(dataJson);
    dataBuf = new Base64URLData(dataJson.getBytes(StandardCharsets.UTF_8));

    generateSignature(keyPair.getPrivateKey());
    messageId = new MessageID(dataBuf, signature);
  }

  public MessageGeneral(
      KeyPair keyPair, Data data, List<PublicKeySignaturePair> witnessSignatures, Gson gson) {
    this(keyPair, data, gson);
    this.witnessSignatures = witnessSignatures;
  }

  private void generateSignature(PrivateKey signer) {
    try {
      signature = signer.sign(dataBuf);
    } catch (GeneralSecurityException e) {
      Timber.tag(TAG).d(e, "failed to generate signature");
    }
  }

  public MessageID getMessageId() {
    return messageId;
  }

  public PublicKey getSender() {
    return sender;
  }

  public Signature getSignature() {
    return signature;
  }

  public List<PublicKeySignaturePair> getWitnessSignatures() {
    return new ArrayList<>(witnessSignatures);
  }

  public Data getData() {
    return data;
  }

  public Base64URLData getDataEncoded() {
    return dataBuf;
  }

  public boolean verify() {
    if (!sender.verify(signature, dataBuf)) {
      return false;
    }

    if (data instanceof WitnessMessageSignature) {
      WitnessMessageSignature witness = (WitnessMessageSignature) data;

      Signature witnessSignature = witness.getSignature();
      MessageID messageID = witness.getMessageId();

      return sender.verify(witnessSignature, messageID);
    }

    return true;
  }

  @NonNull
  @Override
  public String toString() {
    return "MessageGeneral{"
        + "sender='"
        + getSender()
        + '\''
        + ", data='"
        + getData()
        + "', signature='"
        + getSignature()
        + '\''
        + ", messageId='"
        + getMessageId()
        + '\''
        + ", witnessSignatures='"
        + Arrays.toString(witnessSignatures.toArray())
        + "'}";
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/MessageGeneral.kt
package com.github.dedis.popstellar.model.network.method.message

import com.github.dedis.popstellar.model.Immutable
import com.github.dedis.popstellar.model.network.method.message.data.Data
import com.github.dedis.popstellar.model.network.method.message.data.message.WitnessMessageSignature
import com.github.dedis.popstellar.model.objects.security.Base64URLData
import com.github.dedis.popstellar.model.objects.security.KeyPair
import com.github.dedis.popstellar.model.objects.security.MessageID
import com.github.dedis.popstellar.model.objects.security.PrivateKey
import com.github.dedis.popstellar.model.objects.security.PublicKey
import com.github.dedis.popstellar.model.objects.security.Signature
import com.google.gson.Gson
import java.nio.charset.StandardCharsets
import java.security.GeneralSecurityException
import java.util.Objects
import timber.log.Timber

/**
 * Container of a high level message.
 *
 * It is encapsulated inside low level messages
 */
@Immutable
class MessageGeneral {
  val sender: PublicKey
  val dataEncoded: Base64URLData
  val data: Data
  val messageId: MessageID
  val signature: Signature

  var witnessSignatures: List<PublicKeySignaturePair> = ArrayList()
    get() = ArrayList(field)
    private set

  var isEmpty: Boolean = false

  constructor(
      sender: PublicKey,
      dataBuf: Base64URLData,
      data: Data,
      signature: Signature,
      messageID: MessageID,
      witnessSignatures: List<PublicKeySignaturePair>
  ) {
    this.sender = sender
    this.dataEncoded = dataBuf
    this.data = data
    this.messageId = messageID
    this.signature = signature
    this.witnessSignatures = ArrayList(witnessSignatures)
  }

  constructor(keyPair: KeyPair, data: Data?, gson: Gson) {
    requireNotNull(data)

    this.sender = keyPair.publicKey
    this.data = data

    val dataJson = gson.toJson(data, Data::class.java)
    Timber.tag(TAG).d(dataJson)

    this.dataEncoded = Base64URLData(dataJson.toByteArray(StandardCharsets.UTF_8))
    this.signature = generateSignature(keyPair.privateKey)
    this.messageId = MessageID(dataEncoded, signature)
  }

  constructor(
      keyPair: KeyPair,
      data: Data,
      witnessSignatures: List<PublicKeySignaturePair>,
      gson: Gson
  ) : this(keyPair, data, gson) {
    this.witnessSignatures = witnessSignatures
  }

  private fun generateSignature(signer: PrivateKey): Signature {
    return try {
      signer.sign(dataEncoded)
    } catch (e: GeneralSecurityException) {
      Timber.tag(TAG).d(e, "Failed to generate signature")
      Signature("")
    }
  }

  fun verify(): Boolean {
    if (!sender.verify(signature, dataEncoded)) {
      return false
    }

    if (data is WitnessMessageSignature) {
      val witness = data
      val witnessSignature = witness.signature
      val messageID = witness.messageId
      return sender.verify(witnessSignature, messageID)
    }

    return true
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val that = other as MessageGeneral
    return sender == that.sender &&
        dataEncoded == that.dataEncoded &&
        data == that.data &&
        messageId == that.messageId &&
        signature == that.signature &&
        witnessSignatures == that.witnessSignatures
  }

  override fun hashCode(): Int {
    return Objects.hash(sender, dataEncoded, data, messageId, signature, witnessSignatures)
  }

  override fun toString(): String {
    return "MessageGeneral{sender='$sender', data='$data', signature='$signature', messageId='$messageId', witnessSignatures='${
      witnessSignatures.toTypedArray().contentToString()
    }'}"
  }

  companion object {
    private val TAG = MessageGeneral::class.java.simpleName
    val EMPTY = Object()
  }
}
